File: environment/src/main/java/jetbrains/exodus/log/BlockSet.java
/*
 * Copyright 2010 - 2023 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.log;

import jetbrains.exodus.core.dataStructures.hash.LongIterator;
import jetbrains.exodus.core.dataStructures.persistent.PersistentBitTreeLongMap;
import jetbrains.exodus.core.dataStructures.persistent.PersistentLongMap;
import jetbrains.exodus.io.Block;

import java.util.Iterator;

public abstract class BlockSet {
    protected final long blockSize;
    protected final PersistentLongMap<Block> set;

    protected BlockSet(long blockSize, PersistentLongMap<Block> set) {
        this.blockSize = blockSize;
        this.set = set;
    }


    abstract protected PersistentLongMap.ImmutableMap<Block> getCurrent();

    public int size() {
        return getCurrent().size();
    }

    public boolean isEmpty() {
        return getCurrent().isEmpty();
    }

    public Long getMinimum() {
        var iterator = getCurrent().iterator();
        if (iterator.hasNext())
            return iterator.next().getKey() * blockSize;
        return null;
    }

    public Long getMaximum() {
        var iterator = getCurrent().reverseIterator();
        if (iterator.hasNext()) {
            return iterator.next().getKey() * blockSize;
        }
        return null;
    }

    public long[] array() {
        return getFiles(true);
    }

    long[] getFiles() {
        return getFiles(false);
    }

    long[] getFiles(boolean reversed) {
        var current = getCurrent();
        var result = new long[current.size()];
        Iterator<PersistentLongMap.Entry<Block>> iterator;
        if (reversed) {
            iterator = current.reverseIterator();
        } else {
            iterator = current.iterator();
        }
        for (var i = 0; i < result.length; i++) {
            result[i] = iterator.next().getKey() * blockSize;
        }

        return result;
    }

    public boolean contains(long blockAddress) {
        return getCurrent().containsKey(blockAddress / blockSize);
    }

    public Block getBlock(long blockAddress) {
        var result = getCurrent().get(blockAddress / blockSize);
        if (result != null) {
            return result;
        }
        return new EmptyBlock(blockAddress);
    }

    // if address is inside of a block, the block containing it must be included as well if present
    LongIterator getFilesFrom(long blockAddress) {
        return new LongIterator() {
            private final Iterator<PersistentLongMap.Entry<Block>> it;

            {

                if (blockAddress == 0L) {
                    it = getCurrent().iterator();
                } else {
                    it = getCurrent().tailEntryIterator(blockAddress / blockSize);
                }
            }

            @Override
            public long nextLong() {
                return it.next().getKey() * blockSize;
            }

            @Override
            public boolean hasNext() {
                return it.hasNext();
            }

            @Override
            public Long next() {
                return nextLong();
            }

            @Override
            public void remove() {
                throw new UnsupportedOperationException();
            }
        };
    }

    public Mutable beginWrite() {
        return new Mutable(blockSize, set.getClone());
    }


    public static final class Immutable extends BlockSet {
        private final PersistentLongMap.ImmutableMap<Block> immutable;

        public Immutable(long blockSize) {
            this(blockSize, new PersistentBitTreeLongMap<>());
        }

        public Immutable(long blockSize, PersistentLongMap<Block> map) {
            super(blockSize, map);
            this.immutable = map.beginWrite();
        }

        @Override
        protected PersistentLongMap.ImmutableMap<Block> getCurrent() {
            return immutable;
        }
    }

    public static final class Mutable extends BlockSet {
        private final PersistentLongMap.MutableMap<Block> mutable;


        public Mutable(long blockSize, PersistentLongMap<Block> map) {
            super(blockSize, map);
            this.mutable = map.beginWrite();
        }


        @Override
        protected PersistentLongMap.ImmutableMap<Block> getCurrent() {
            return mutable;
        }

        public void clear() {
            mutable.clear();
        }

        public void add(long blockAddress, Block block) {
            mutable.put(blockAddress / blockSize, block);
        }

        public boolean remove(long blockAddress) {
            return mutable.remove(blockAddress / blockSize) != null;
        }

        public Immutable endWrite() {
            if (!mutable.endWrite()) {
                throw new IllegalStateException("File set can't be updated");
            }
            return new Immutable(blockSize, set.getClone());
        }
    }

    @SuppressWarnings("ClassCanBeRecord")
    private static final class EmptyBlock implements Block {
        private final long address;

        private EmptyBlock(long address) {
            this.address = address;
        }

        @Override
        public long getAddress() {
            return address;
        }

        @Override
        public long length() {
            return 0L;
        }

        @Override
        public int read(byte[] output, long position, int offset, int count) {
            return 0;
        }

        @Override
        public Block refresh() {
            return this;
        }
    }
}

File: environment/src/main/kotlin/jetbrains/exodus/log/BlockSet.kt
/*
 * Copyright 2010 - 2023 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.log

import jetbrains.exodus.core.dataStructures.hash.LongIterator
import jetbrains.exodus.core.dataStructures.persistent.PersistentBitTreeLongMap
import jetbrains.exodus.core.dataStructures.persistent.PersistentLongMap
import jetbrains.exodus.io.Block

// block key is aligned block address, i.e. block address divided by blockSize
sealed class BlockSet(@JvmField val blockSize: Long, @JvmField val set: PersistentLongMap<Block>) {
    protected abstract fun getCurrent(): PersistentLongMap.ImmutableMap<Block>

    fun size() = getCurrent().size()

    fun isEmpty() = getCurrent().isEmpty

    fun getMinimum(): Long? {
        val iterator = getCurrent().iterator()

        if (iterator.hasNext()) {
            return iterator.next().key.keyToAddress()
        }

        return null
    }

    fun getMaximum(): Long? {
        val iterator = getCurrent().reverseIterator()
        if (iterator.hasNext()) {
            return iterator.next().key.keyToAddress()
        }

        return null
    }

    fun getFiles(reversed: Boolean = false): LongArray {
        val current = getCurrent()
        val result = LongArray(current.size())
        val it = if (reversed) {
            current.reverseIterator()
        } else {
            current.iterator()
        }
        for (i in 0 until result.size) {
            result[i] = it.next().key.keyToAddress()
        }
        return result
    }

    fun contains(blockAddress: Long) = getCurrent().containsKey(blockAddress.addressToKey())

    fun getBlock(blockAddress: Long): Block =
        getCurrent().get(blockAddress.addressToKey()) ?: EmptyBlock(blockAddress)

    // if address is inside of a block, the block containing it must be included as well if present
    fun getFilesFrom(blockAddress: Long = 0L): LongIterator = object : LongIterator {
        val current = getCurrent()
        val it =
            if (blockAddress == 0L) current.iterator() else current.tailEntryIterator(blockAddress.addressToKey())

        override fun next() = nextLong()

        override fun hasNext() = it.hasNext()

        override fun nextLong() = it.next().key.keyToAddress()

        override fun remove() = throw UnsupportedOperationException()
    }

    fun beginWrite() = Mutable(blockSize, set.clone)


    protected fun Long.keyToAddress(): Long = this * blockSize

    protected fun Long.addressToKey(): Long = this / blockSize

    class Immutable(
        blockSize: Long,
        map: PersistentLongMap<Block> = PersistentBitTreeLongMap()
    ) : BlockSet(blockSize, map) {
        private val immutable: PersistentLongMap.ImmutableMap<Block> = map.beginRead()

        public override fun getCurrent(): PersistentLongMap.ImmutableMap<Block> = immutable
    }

    class Mutable(blockSize: Long, map: PersistentLongMap<Block>) : BlockSet(blockSize, map) {
        private val mutable: PersistentLongMap.MutableMap<Block> = set.beginWrite()

        override fun getCurrent(): PersistentLongMap.ImmutableMap<Block> = mutable

        fun clear() = mutable.clear()

        fun add(blockAddress: Long, block: Block) = mutable.put(blockAddress.addressToKey(), block)

        fun remove(blockAddress: Long) = mutable.remove(blockAddress.addressToKey()) != null

        fun endWrite(): Immutable {
            if (!mutable.endWrite()) {
                throw IllegalStateException("File set can't be updated")
            }
            return Immutable(blockSize, set.clone)
        }
    }

    private inner class EmptyBlock(private val address: Long) : Block {
        override fun getAddress() = address

        override fun length() = 0L

        override fun read(output: ByteArray, position: Long, offset: Int, count: Int) = 0

        override fun refresh() = this
    }
}
